### PR TITLE
TensorRT EP: Weightless API integration in ONNX Runtime

### PR DESCRIPTION
**Pull request by moraxu**
TensorRT EP: Weightless API integration in ONNX Runtime

### CODE REVIEW

File: onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.cc [hunk 1]
```diff
@@ -8,8 +8,10 @@
 #include "onnx_ctx_model_helper.h"
 #include "core/providers/cuda/shared_inc/cuda_call.h"
 #include "core/framework/execution_provider.h"
+#include "tensorrt_execution_provider.h"
```
Comment (moraxu) [inline onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.cc:11]: The header is in the same dir so not sure about the warning?

File: onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.h [hunk 2]
```diff
@@ -15,6 +15,7 @@ static const std::string EPCONTEXT_OP = "EPContext";
 static const std::string EMBED_MODE = "embed_mode";
 static const std::string EP_CACHE_CONTEXT = "ep_cache_context";
 static const std::string COMPUTE_CAPABILITY = "hardware_architecture";
+static const std::string ONNX_MODEL_FILENAME = "onnx_model_filename";
```
Comment (moraxu) [inline onnxruntime/core/providers/tensorrt/onnx_ctx_model_helper.h:19]: Will ignore the warning for the sake of consistency?

File: include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h [hunk 3]
```diff
@@ -64,10 +64,20 @@ struct OrtTensorRTProviderOptionsV2 {
...
+   *
    */
   int trt_dump_ep_context_model{0};               // Dump EP context node model
   const char* trt_ep_context_file_path{nullptr};  // Specify file name to dump EP context node model. Can be a path or a file name or a file name with path.
   int trt_ep_context_embed_mode{0};               // Specify EP context embed mode. Default 0 = context is engine cache path, 1 = context is engine binary data
+  int trt_weightless_engine_enable{0};            // Enable weightless engine build. Default 0 = false,
+                                                  // nonzero = true
+  const char* trt_onnx_model_folder_path{nullptr};  // Folder path relative to the current working directory for
+                                                    // the ONNX model containing the weights (applicable only when
+                                                    // the "trt_weightless_engine_enable" is enabled)
```
Comment (kevinch-nv) [inline include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:80]:
```suggestion
                                                    // the "trt_weightless_engine_enable" option is enabled)
```

File: onnxruntime/core/graph/contrib_ops/contrib_defs.cc [hunk 4]
```diff
@@ -3233,11 +3233,16 @@ void RegisterContribSchemas() {
           "The path is relative to this Onnx file. Default is 1.",
           AttributeProto::INT,
           static_cast<int64_t>(1))
-      .Attr(
+              .Attr(
```
Comment (kevinch-nv) [inline onnxruntime/core/graph/contrib_ops/contrib_defs.cc:3236]: Looks like bad formatting?

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 5]
```diff
@@ -1330,6 +1332,12 @@ TensorrtExecutionProvider::TensorrtExecutionProvider(const TensorrtExecutionProv
         engine_cache_enable_ = (std::stoi(engine_cache_enable_env) == 0 ? false : true);
       }
 
+      const std::string weightless_engine_enable_env
+        = onnxruntime::GetEnvironmentVar(tensorrt_env_vars::kWeightlessEngineEnable);
+      if (!weightless_engine_enable_env.empty()) {
+        weightless_engine_enable_ = (std::stoi(weightless_engine_enable_env) == 0 ? false : true);
```
Comment (kevinch-nv) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:1338]:
```suggestion
        weightless_engine_enable_ = std::stoi(weightless_engine_enable_env) != 0;
```

File: include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h [hunk 6]
```diff
@@ -64,10 +64,20 @@ struct OrtTensorRTProviderOptionsV2 {
...
+   * 3. In the case of building weightless engines, the same security reasons as listed in 1) apply to the
+   *    "onnx_model_filename" node attribute of EP context node, which contains a filename of the ONNX model with the
+   *    weights needed for the refit process. User can specify a folder path relative to the current working
+   *    directory by means of the "trt_onnx_model_folder_path" option.
+   *
    */
   int trt_dump_ep_context_model{0};               // Dump EP context node model
   const char* trt_ep_context_file_path{nullptr};  // Specify file name to dump EP context node model. Can be a path or a file name or a file name with path.
   int trt_ep_context_embed_mode{0};               // Specify EP context embed mode. Default 0 = context is engine cache path, 1 = context is engine binary data
+  int trt_weightless_engine_enable{0};            // Enable weightless engine build. Default 0 = false,
```
Comment (kevinch-nv) [inline include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:76]: Is the user able to serialize the full engine after refitting? After reading through this MR it doesn't seem like that's possible.

File: onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc [hunk 7]
```diff
@@ -3243,6 +3268,14 @@ Status TensorrtExecutionProvider::CreateNodeComputeInfoFromGraph(const GraphView
       if (trt_state->auxiliary_streams >= 0) {
         LOGS_DEFAULT(WARNING) << "[TensorRT EP] Auxiliary streams can only be set on TRT 8.6 onwards!";
       }
+#endif
```
Comment (kevinch-nv) [inline onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:3271]: Looks like these lines are identical to 2765. Can this be refactored into a common function?

Comment (moraxu) [reply onnxruntime/core/providers/tensorrt/tensorrt_execution_provider.cc:3271]: Hmm, those logging macros eventually use `ORT_WHERE` which pulls in `__LINE__` and `__FUNCTION__`

File: include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h (hunk 6 above)
Comment (moraxu) [reply include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:76]: Not for now, as discussed with Max, we'd like the user being able to always choose the same session options, i.e. ultimately choosing between the `model_ctx.onnx` (if present) and `model.onnx` (using it first once in the absence of the former). The serialized engine with refitted weights would most likely introduce a yet another EP option to the flow.

File: include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h [hunk 8]
```diff
@@ -64,10 +64,20 @@ struct OrtTensorRTProviderOptionsV2 {
...
+                                                  // nonzero = true
+  const char* trt_onnx_model_folder_path{nullptr};  // Folder path relative to the current working directory for
```
Comment (chilo-ms) [inline include/onnxruntime/core/providers/tensorrt/tensorrt_provider_options.h:78]: This struct member could be pointing to a memory location allocated by TRT EP [here](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/providers/tensorrt/tensorrt_execution_provider_info.cc#L251).
Therefore, we should add code to release it [here](https://github.com/microsoft/onnxruntime/blob/main/onnxruntime/core/session/provider_bridge_ort.cc#L2238).

**Pull request merged**